RandLAPACK/comps/rl_preconditioners.hh

```diff
@@ -0,0 +1,208 @@
+ * We do not perform this scaling ourselves, since other values
+ * of mu might be intended.
+ * 
+ * @param[in] layout_A
+ *      Either blas::Layout::RowMajor or blas::Layout::ColMajor.
+ * @param[in] m
+ *      The number of rows in A
+ * @param[in] n
+ *      The number of columns in A (and rows in M)
+ * @param[in] d
+ *      The number of rows in the sketched data matrix; must be >= n.
+ * @param[in] k
+ *      The number of nonzeros per column in the sketching operator
+ *      A common value is k=8 when n is on the order of a couple thousand.
+ * @param[in] A
+ *      A buffer for an m \times n matrix A. Interpreted in 
+ *      "layout_A" order.
+ * @param[out] V_sk
+ *      We require that V_sk.size() >= d*n. Letting "r" denote the return
+ *      value of this function, on exit the orthogonal factor of the preconditioner
+ *      is identified with the first n*r entries in V_sk.data(),
+ *      read in column-major format.
+ * @param[in] sigma_sk
+ *      We require sigma_sk.size() >= n.
+ * @param[in] threads
+ *      The number of OpenMP threads used during sketching.
+ * @param[in] seed_key
+ *      The key provided to an underlying counter-based random number generator.
+ * @param[inout] seed_ctr
+ *      A 32-bit counter provided to the underlying counter-based random number
+ *      generator. This value is incremented by m*k on-exit.
+ * @returns
+ *      void
+ */
+template <typename T>
+void rpc_data_svd_saso(
+    blas::Layout layout_A,
+    int64_t m, // number of rows in A
+    int64_t n, // number of columns in A
+    int64_t d, // number of rows in sketch of A
+    int64_t k, // number of nonzeros in each column of the sketching operator
+    std::vector<T>& A, // buffer of size m*n.
+    std::vector<T>& V_sk, // length at least d*n 
+    std::vector<T>& sigma_sk, // length at least n
+    int64_t threads, // number of OpenMP threads to use in sketching.
+    uint64_t seed_key,
+    uint32_t seed_ctr
+) {
+    T* buff_A = A.data();
+    if (V_sk.size() < d*n)
+        throw std::invalid_argument("V_sk must be of size at least d*n.");
+    T* buff_A_sk = V_sk.data();
+    
+    // step 1.1: define the sketching operator
+    RandBLAS::sparse::SparseDist D{
+        .n_rows = d,
+        .n_cols = m,
+        .vec_nnz=k
+    };
+    auto state = RandBLAS::base::RNGState(seed_key, seed_ctr);
+    RandBLAS::sparse::SparseSkOp<T> S(D, state);
+    RandBLAS::sparse::fill_sparse(S);
+
+    // step 1.2: sketch the data matrix
+    int64_t lda, lda_sk;
+    if (layout_A == blas::Layout::RowMajor) {
+        lda = n;
+        lda_sk = n;
+    } else {
+        lda = m;
+        lda_sk = d;
+    }
+    blas::scal(d*n, 0.0, buff_A_sk, 1);
+    RandBLAS::sparse::lskges<T>(
+        layout_A,
+        blas::Op::NoTrans,
+        blas::Op::NoTrans,
+        d, n, m,
+        1.0, S, 0, 0,
+        buff_A, lda,
+        0.0, buff_A_sk, lda_sk
+    );
+
+    // step 2: compute SVD of sketch
+    T* ignore;
```

Comment:
Explicitely initializing this to NULL might be nice so when debugging code seeing the NULL pointers below looks intended where as we might just get a pointer to a strange part of memory that has nonsensical data otherwise

Reply:
Fixed.